### PR TITLE
Consider running GC when allocating and synchronizing

### PR DESCRIPTION
**Pull request by maleadt**
Implements https://github.com/JuliaGPU/CUDA.jl/issues/2303; cc @gbaraldi

### CODE REVIEW

File: src/pool.jl [hunk 1]
```diff
@@ -81,10 +72,73 @@ function actual_free(buf::Mem.DeviceBuffer;
...
+  end
+  stats.last_time[] = current_time
+
+  # finally, call the GC
+  pre_gc_live = stats.live[]
+  gc_time = @elapsed GC.gc()
+  post_gc_live = stats.live[]
+  Threads.atomic_set!(stats.last_freed, memory_freed)
+  ## GC times can vary, so smooth them out
+  Threads.atomic_set!(stats.last_gc_time, 0.75*stats.last_gc_time[] + 0.25*gc_time)
```
Comment (maleadt) [inline src/pool.jl]: This is me going "now what would be a good GC heuristic" fully ignoring decades of research on the topic. So @gbaraldi if you have any thoughts; feel free.

File: src/pool.jl [hunk 2]
```diff
@@ -6,36 +6,27 @@ using Logging
...
-
 mutable struct AllocStats
   alloc_count::Threads.Atomic{Int}
   alloc_bytes::Threads.Atomic{Int}
 
   free_count::Threads.Atomic{Int}
   free_bytes::Threads.Atomic{Int}
 
-  total_time::MaybeAtomicFloat64
+  total_time::Threads.Atomic{Float64}
```
Comment (vchuravy) [inline src/pool.jl:16]: Can we use `@atomic total_time::Float64`?

Comment (maleadt) [reply src/pool.jl:16]: I guess we could update almost all of these then?

File: src/pool.jl (hunk 1 above)
Comment (maleadt) [reply src/pool.jl]: @DhairyaLGandhi reported that this didn't help his memory-hungry application, so the heuristics probably need tuning.

**Comment by BioTurboNick** [comment]
Trying this PR on my example from https://github.com/FluxML/Flux.jl/issues/2414 on my NVIDIA RTX 4080 on Windows 11

... and it hung Julia and made the process unkillable in Task Manager.

EDIT: apparently this is often caused by a driver fault. https://superuser.com/questions/136272/how-can-i-kill-an-unkillable-process. And eventually I got a BSOD for KMODE_EXCEPTION_NOT_HANDLED in the NVIDIA driver

EDIT2: Though also this is in comparison with 5.2, not the current master. ... and the base of the branch hung my whole system 🙃

**Comment by CarloLucibello** [comment]
The example in https://discourse.julialang.org/t/gpu-memory-usage-increasing-on-each-epoch-flux/112942/2
is worth benchmarking as well. Let me know if you need help with testing!

**Comment by maleadt** [comment]
Thanks, the example from https://discourse.julialang.org/t/gpu-memory-usage-increasing-on-each-epoch-flux/112942 was useful. I tuned the heuristic, and in low-memory situations it significantly improves performance, while with more reasonable amounts of memory available it smoothens out the cost of garbage collection and results in slightly shorter pauses and more consistent execution times. It's possible to disable the heuristic by setting `JULIA_CUDA_GC_EARLY=false`, making it easy to compare.

Taking the Flux example from Discourse:

```julia
using Flux
using MLUtils: DataLoader
using CUDA
using NVTX

function increasing_gpu_memory_usage()
    n_obs = 300_000
    n_feature = 1000
    X = rand(n_feature, n_obs)
    y = rand(1, n_obs)
    train_data = DataLoader((X, y) |< gpu; batchsize = 2048, shuffle = false)

    model = Dense(n_feature, >) |< gpu
    loss(m, _x, _y) = Flux.Losses.mse(m(_x), _>)
    opt_state = Flux.setup(Flux.Adam(), model)
    Flux.train!(loss, model, train_data, opt_state)
    total_time = @elapsed begin
        CUDA.@profile external=true for epoch in 1:100
            NVTX.@range "Epoch $epoch" begin
                train_time = @elapsed Flux.train!(loss, model, train_data, opt_state)
                @info "Epoch $(epoch) train time $(round(train_time, digits=3))"
            end
        end
    end
    @info "Total time $(round(total_time, digits=3))"
    return
end

isinteractive() || increasing_gpu_memory_usage()
```

Old behavior, with a 4GiB memory limit:

```
❯ JULIA_CUDA_GC_EARLY=false JULIA_CUDA_HARD_MEMORY_LIMIT=4GiB julia --project wip.jl
...
[ Info: Epoch 90 train time 0.031
retry_reclaim: freed 2.865 GiB
[ Info: Epoch 91 train time 0.031
[ Info: Epoch 92 train time 0.027
retry_reclaim: freed 2.865 GiB
[ Info: Epoch 93 train time 0.03
retry_reclaim: freed 2.873 GiB
[ Info: Epoch 94 train time 0.031
retry_reclaim: freed 2.873 GiB
[ Info: Epoch 95 train time 0.03
retry_reclaim: freed 2.873 GiB
[ Info: Epoch 96 train time 0.031
[ Info: Epoch 97 train time 0.027
retry_reclaim: freed 2.873 GiB
[ Info: Epoch 98 train time 0.031
retry_reclaim: freed 2.865 GiB
[ Info: Epoch 99 train time 0.031
retry_reclaim: freed 2.865 GiB
[ Info: Epoch 100 train time 0.031
[ Info: Total time 4.307
```

Note that it takes a while to reach steady-state, so I'm only showing the final epochs.

Enabling the new heuristic:

```
❯ JULIA_CUDA_GC_EARLY=true JULIA_CUDA_HARD_MEMORY_LIMIT=4GiB julia --project wip.jl
...
[ Info: Epoch 90 train time 0.031
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
maybe_collect: collected 1.869 GiB (9.0% < 10.0%) while using 3.004 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 91 train time 0.033
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 92 train time 0.031
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 93 train time 0.031
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 94 train time 0.03
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 95 train time 0.03
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
maybe_collect: collected 1.869 GiB (9.0% < 10.0%) while using 3.004 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 96 train time 0.033
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 97 train time 0.03
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 98 train time 0.03
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 99 train time 0.03
maybe_collect: collected 1.877 GiB (9.0% < 10.0%) while using 3.003 GiB/4.000 GiB = 75.09% (> 75.0%) memory
[ Info: Epoch 100 train time 0.03
[ Info: Total time 3.76
```

Part of the advantage seems comes from the fact that collecting earlier makes it possible for memory to be made available to the memory pool without having to explicity synchronize. Before, we called the GC when we were already at 100% memory usage, and because memory gets freed asynchronously (i.e. it only becomes available when the free actually executes), that meant that we often had to also wait for the GPU to finish its current work. While now, because of collecting earlier, we give the free a chance to materialize obviating explicit synchronization.

Everybody, please test this out on your code, or share (easily reproducible) MWEs that illustrate problems.

**Comment by maleadt** [comment]
I don't consider this ready, but am going to go ahead and merge this to avoid excessive conflicts with the memory refactor I'm doing in https://github.com/JuliaGPU/CUDA.jl/pull/2335.
